cmd/create/externalauthproviders/cmd.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright (c) 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package externalauthproviders
+
+import (
+	"os"
+
+	cmv1 "github.com/openshift-online/ocm-sdk-go/clustersmgmt/v1"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/rosa/pkg/externalauthproviders"
+	"github.com/openshift/rosa/pkg/input"
+	"github.com/openshift/rosa/pkg/interactive"
+	"github.com/openshift/rosa/pkg/ocm"
+	"github.com/openshift/rosa/pkg/rosa"
+)
+
+var externalAuthProvidersArgs *externalauthproviders.ExternalAuthProvidersArgs
+
+const argsPrefix string = ""
+
+var Cmd = &cobra.Command{
+	Use:     "external-auth-provider",
+	Aliases: []string{"externalauthproviders", "externalauthprovider", "external-auth-provider"},
+	Short:   "Create an external authentication provider for a cluster.",
+	Long:    "Configuring the cluster with external authentication configuration instead of internal oidc provider. ",
+	Example: `  # Interactively create an external authentication provider to a cluster named "mycluster"
+  rosa create external-auth-provider --cluster=mycluster --interactive`,
+	Run:    run,
+	Hidden: true,
+	Args:   cobra.NoArgs,
+}
+
+func init() {
+	flags := Cmd.Flags()
+
+	ocm.AddClusterFlag(Cmd)
+	interactive.AddFlag(flags)
+	externalAuthProvidersArgs = externalauthproviders.AddExternalAuthProvidersFlags(Cmd, argsPrefix)
+}
+
+func run(cmd *cobra.Command, _ []string) {
+	r := rosa.NewRuntime().WithOCM()
+	defer r.Cleanup()
+
+	clusterKey := r.GetClusterKey()
+	cluster := r.FetchCluster()
+
+	input.CheckIfHypershiftClusterOrExit(r, cluster)
```

Comment:
@chenz4027 As a suggestion, what if you had an `ExternalAuthService` interface with a couple of methods like this:

```
type ExternalAuthService interface {
  IsExternalAuthProviderSupported(cluster *cmv1.Cluster) error
  CreateExternalAuthProvider(cluster *cmv1.Cluster, args ExternalAuthProviderArgs) error
}
```

You could then move most of your logic from this command into that implementation and unit test it. We could then update this command to look _something_ like:

```
func runWithRuntime(r *rosa.Runtime, cmd *cobra.Command, argv []string) error {

 externalAuthService := NewExternalAuthService() 
 err := externalAuthService.IsExternalAuthAvailable(cluster)
 if err != nil {
   return err
 }

 externalAuthProvidersArgs, err := externalauthproviders.GetExternalAuthOptions(
			cmd.Flags(), "", false, externalAuthProvidersArgs)

 if err != nil {
  return err
 }

 err = externalAuthService.CreateExternalAuthProvider(cluster, externalAuthProvidersArgs)
 if err != nil {
   return err
 }

 r.Reporter.Infof("Successfully created external authentication configuration for cluster '%s'", cluster.ID())
}
```

I see a couple of benefits if we structure in that way:

1. Cobra is just a thin wrapper around our actual logic
2. We can unit test the ExternalAuthService implementation using mocks
3. We can more easily unit test your `runWithRuntime` function as well.

Reply:
@robpblake For sure, will work on refactoring and adding more tests to it. Thanks a lot @robpblake 💯

Reply:
@robpblake I tried a first iteration with a run calling runWithRuntime but got a few issues. I wonder what NewExternalAuthService() would be 🤔

Reply:
Added runWithRuntime, and calls it from run. However, was stuck with the interface and its new service.

Reply:
Thanks to @robpblake for more instructions from slack, refactored to externalAuthService.